### DISCUSSION TITLE
Is there a way to update lambda code concurrently ?

### DISCUSSION THREAD

**Discussion by olup** [pulumi/pulumi#9003] [Pulumi in Practice] [answered]
Hi,

I have been asking on the slack but I think I might have more chances here. I bundle forty or so lambdas with esbuild and then feed each file to a fileAsset, archiveAsset, and eventually a lambda function class instance. This works very well ! Code gets detected as different when it changes and updated on AWS lambda.

However, When the code changes a bit everywhere, 40 update jobs get started - in the console they all start together. But it seems to me they finish sequentially, one after the other one, and that the uploading of the archive is not parallel, resulting in quite the waiting time.

Is there a way to parallelize those updates ? It's the only thing I miss in my use case to make Pulumi the absolute iac tool.

Thanks.

**Comment by Frassle** [comment, marked as answer]
I'm pretty sure the engine will do updates for functions in parallel anyway. Are you sure your code isn't adding a dependency chain between the functions objects or the archives? Do you have any log output from `pulumi up` showing this behavior? (you can run with `--logtostderr -v9` for more verbose debug logs)

**Reply by olup** [reply]
Ok, that's a lot of logs. What should I look for specifically ?

**Reply by Frassle** [reply]
Log lines for "ResourceMonitor.RegisterResource" and see what "deps" is set to.

**Reply by olup** [reply]
MMh most of them only depends on a role and a layer : 
`I0217 13:06:18.463977   38056 source_eval.go:1055] ResourceMonitor.RegisterResource received: t=aws:lambda/function:Function, name=prj-serverless-dev-update-store-status, custom=true, #props=9, parent=urn:pulumi:dev::prj-serverless::pulumi:pulumi:Stack::prj-serverless-dev, protect=false, provider=urn:pulumi:dev::prj-serverless::pulumi:providers:aws::default_4_37_2::f64936d0-6f40-46bd-ad96-30f7cb207a12, deps=[urn:pulumi:dev::prj-serverless::aws:lambda/layerVersion:LayerVersion::prj-serverless-dev-prisma-layer urn:pulumi:dev::prj-serverless::aws:iam/role:Role::prj-serverless-dev-lambda-role], deleteBeforeReplace=<nil>, ignoreChanges=[], aliases=[], customTimeouts={0 0 0}, providers=map[], replaceOnChanges=[]
`

**Reply by olup** [reply]
(thanks for bearing with me btw)

**Reply by Frassle** [reply]
Is it the LayerVersion objects or the Function objects that don't appear to be updated in parallel?

Do the RegisterResourceCalls happen close together? Can you see "Provider[aws, 0x****].Update" log lines, do they happen close together? Do the "executing" and "success" lines happen in order?

I'll ask some of our aws team if they have any ideas about this as well.

**Reply by olup** [reply]
Many thanks. In the problem I am facing, only the functions object's codes don't appear to be updated in parallel. The layer hasn't changed and so is not updated. 

I will try to get an answer for the closeness of the call, but it's hard to say from reading the logs as "close" is not an absolute metric

**Reply by Frassle** [reply]
Ah I've found why! 

Your using our terraform based aws provider and the underlying terraform code has a global lock around function code uploads. See https://github.com/hashicorp/terraform-provider-aws/blob/9c0b16cae61bf48eb76a19e089f823101600b4fc/internal/service/lambda/function.go#L1160:
```
			// Grab an exclusive lock so that we're only reading one function into
			// memory at a time.
			// See https://github.com/hashicorp/terraform/issues/9364
			conns.GlobalMutexKV.Lock(awsMutexLambdaKey)
			defer conns.GlobalMutexKV.Unlock(awsMutexLambdaKey)
```

I don't think our native aws provider does this, although that may mean we hit the problem described in https://github.com/hashicorp/terraform/issues/9364 as we load GBs of code at once.

**Reply by olup** [reply]
Interresting. How can I avoid the terraform adapter ?

**Reply by olup** [reply]
I tryed to switch to oulumi aws native, but the code part expects a string, which is fine, but then thows when uploading : `creating resource: operation error CloudControl: CreateResource, https response error StatusCode: 400, RequestID: 7b54c2cd-4fc3-4937-ac21-1173ff6132ba, api error ValidationException: 1 validation error detected: Value at 'desiredState' failed to satisfy constraint: Member must have length less than or equal to 65536` I guess code is too heavy to be updated this way

**Reply by olup** [reply]
Would it be the same if I first uploaded the code to s3 before deploying my lambdas and pointing to the bucket's object ?

**Reply by Frassle** [reply]
It doesn't look like terraform does any mutual exclusion for S3 objects, so that will probably run fully in parallel.

**Reply by olup** [reply]
Cheers! So here's what I did: creating an s3 object with my computed archive, and using a sha1 of the file content as the s3 object key to trigger function update when asset changes (function links to s3 bucket and key). And this has been working quite well, and is very fast. Thanks a lot for landing a hand here @Frassle :-)

**Comment by nimbinatus** [comment]
Note that GitHub Discussions doesn't let us mark a thread response as an answer, so we're marking the top of the thread that contains the answer 😄
